tests/trainer/test_trainer.py

```diff
@@ -192,6 +207,56 @@ def _optimizer_step(epoch, batch_idx, optimizer, optimizer_idx,
     trainer.fit(model)
 
 
+@pytest.mark.parametrize(
+    ['schedule', 'limit'],
+    [
+        pytest.param({1: 2, 3: 4}, 1.0),
+        pytest.param({1: 2, 3: 4}, 0.5),  # not to be divisible by accumulate_grad_batches on purpose
+        pytest.param(3, 1.0),
+        pytest.param(3, 0.8),  # not to be divisible by accumulate_grad_batches on purpose
+        pytest.param(4, 1.0),
+        pytest.param(4, 0.7),  # not to be divisible by accumulate_grad_batches on purpose
+    ],
+)
+def test_gradient_accumulation_scheduling_last_batch(tmpdir, schedule, limit):
+    """
+    Verify optimizer.step() applied to last batch while grad accumulation
+    """
+
+    model = EvalModelTemplate()
+
+    trainer = Trainer(
+        accumulate_grad_batches=schedule,
+        max_epochs=4,
+        limit_train_batches=limit,
+        default_root_dir=tmpdir
+    )
+
+    loss_backward = [0] * len(list(model.parameters()))
+    opt_step = [torch.tensor([0.0])] * len(list(model.parameters()))
+
+    def on_after_backward_():
+        for i, param in enumerate(model.parameters()):
+            loss_backward[i] = param.clone().data
+
+    def on_before_zero_grad_(optimizer):
+        for i, param in enumerate(model.parameters()):
+            opt_step[i] = param.clone().data
+
+    def on_train_batch_end_(batch, batch_idx, dataloader_idx):
+        # to check optimizer.step() get called or not
+        for i, (loss_param, opt_param) in enumerate(zip(loss_backward, opt_step)):
+            if trainer.num_training_batches == batch_idx + 1:
+                # model.parameters() get updated
+                assert torch.equal(loss_param, opt_param) is False
+
+    # for the test
+    model.on_after_backward = on_after_backward_
+    model.on_before_zero_grad = on_before_zero_grad_
+    model.on_train_batch_end = on_train_batch_end_
```

Comment:
Can you structure this by creating a custom model `CurrentModel(EvalModelTemplate)` and override these methods? You can access the trainer using `self.trainer`.

Reply:
Alright, Got it.

Reply:
how can i access `model.parameters()`?

Reply:
self.parameters()

Reply:
or better use `self.state_dict()`. You can avoid this for loop with state_dict
```python
for i, param in enumerate(model.parameters()):
    opt_step[i] = param.clone().data
```
Something similar to this: https://github.com/PyTorchLightning/pytorch-lightning/blob/6c18fd9a24585c89dff888ef31facfb89c7fe0c2/tests/trainer/test_lr_finder.py#L37-L45

Reply:
Ohh yeah, i forgot to call `super()` 😅